### PR TITLE
Case attachment refactor

### PR DESCRIPTION
**Pull request by millerdev**
Drop columns `attachment_src` and `attachment_from`. Merge `identifier` with `name` (both had the same value). Another PR will follow to drop the unused columns and database functions.

@snopoke @jmtroth0

### CODE REVIEW

File: corehq/form_processor/models.py [hunk 1]
```diff
@@ -984,28 +982,16 @@ def from_form_attachment(self, attachment):
...
 
-        if not self.content_type and self.attachment_src:
-            guessed = mimetypes.guess_type(self.attachment_src)
+        if not self.content_type and attachment_src:
+            guessed = mimetypes.guess_type(attachment_src)
             if len(guessed) > 0 and guessed[0] is not None:
                 self.content_type = guessed[0]
 
-        if isinstance(attachment, CaseAttachmentSQL):
-            assert self.identifier == attachment.identifier
```
Comment (snopoke) [inline corehq/form_processor/models.py]: shouldn't this assert remain as `assert self.name == attachment.name`?

File: corehq/ex-submodules/casexml/apps/case/templates/case/partials/case_attachments.html [hunk 2]
```diff
@@ -6,8 +6,6 @@
     {% if not forloop.first %}<br>{% endif %}
     <dt>{% trans "Case Property" %}</dt>
         <dd>{{ attach_key }}</dd>
-    <dt>{% trans "Source File" %}</dt>
-        <dd>{{ attachment.attachment_src }}</dd>
     <dt>{% trans "File Type" %}</dt>
         <dd>{{ attachment.server_mime }}</dd>
```
Comment (snopoke) [inline corehq/ex-submodules/casexml/apps/case/templates/case/partials/case_attachments.html]: Just seeing this now and unrelated but this should be `attachment.content_type`

File: corehq/ex-submodules/casexml/apps/case/xml/parser.py [hunk 3]
```diff
@@ -191,18 +189,12 @@ def from_v2(cls, block):
...
             if isinstance(data, six.string_types):
-                attachment_from = None
                 attachment_src = None
                 attachment_name = None
             else:
-                attachment_from = data.get('@from', None)
                 attachment_src = data.get('@src', None)
                 attachment_name = data.get('@name', None)
-
-            if attachment_from == attachment_src == attachment_name == None:
```
[the remark this answers is not in the document]
Comment (millerdev) [reply corehq/ex-submodules/casexml/apps/case/xml/parser.py]: Is this `if` block just for documentation purposes? It doesn't do anything, right?

File: corehq/form_processor/abstract_models.py [hunk 4]
```diff
@@ -288,12 +288,10 @@ def get_properties_in_api_format(self):
...
     @memoized
     def get_attachment_map(self):
-        return dict([
-            (name, {
-                'url': self.get_attachment_server_url(att.identifier),
-                'mime': att.attachment_from
-            }) for name, att in self.case_attachments.items()
-        ])
+        return {
+            name: {'url': self.get_attachment_server_url(att.identifier)}
```
[the remark this answers is not in the document]
Comment (millerdev) [reply corehq/form_processor/abstract_models.py]: Sure. But what is "mime" here? Is it related to [MIME](https://en.wikipedia.org/wiki/MIME)? As you previously pointed out, the value was always `"local"`.

File: corehq/form_processor/migrations/0071_case_attachment_refactor.py [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+class Migration(migrations.Migration):
+
+    dependencies = [
+        ('form_processor', '0070_auto_20180329_1016'),
+    ]
+
+    operations = [
+        migrations.AlterIndexTogether(
+            name='caseattachmentsql',
+            index_together=set([('case', 'name')]),
```
Comment (emord) [inline corehq/form_processor/migrations/0071_case_attachment_refactor.py]: This will lock the table when creating the index since it isn't concurrent by default. May want to see how large the table is on each env to see how long this will take. Also I don't see this dropping the index on (case, identifier)

File: corehq/form_processor/abstract_models.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (millerdev) [reply corehq/form_processor/abstract_models.py]: Is `content_type` referenced anywhere? If not, why include it?

File: corehq/form_processor/models.py (hunk 1 above)
Comment (millerdev) [reply corehq/form_processor/models.py]: I don't think this method is ever called with `CaseAttachmentSQL`, so I don't think this assertion is ever executed.

File: corehq/form_processor/abstract_models.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (snopoke) [reply corehq/form_processor/abstract_models.py]: this is what gets output in the Case API.

File: corehq/form_processor/migrations/0071_case_attachment_refactor.py (hunk 5 above)
Comment (millerdev) [reply corehq/form_processor/migrations/0071_case_attachment_refactor.py]: There are very few projects that use case attachments, and therefore the table is not very big. I just looked and it's less than 4K rows per shard.

The (case, identifier) index will be deleted when the `identifier` column is dropped by #21107

File: corehq/ex-submodules/casexml/apps/case/xml/parser.py [hunk 6]
```diff
@@ -159,16 +159,14 @@ class CaseAttachment(object):
     A class that wraps an attachment to a case
     """
 
-    def __init__(self, identifier, attachment_src, attachment_from, attachment_name):
```
[the remark this answers is not in the document]
Comment (millerdev) [reply corehq/ex-submodules/casexml/apps/case/xml/parser.py]: d004a43f72c62587c82909604fdf1f5d40a0bb99

File: corehq/form_processor/abstract_models.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (millerdev) [reply corehq/form_processor/abstract_models.py]: 30b81c0f5976b5523f1f6db1afd6745ef5186c2a

File: corehq/form_processor/abstract_models.py [hunk 7]
```diff
@@ -354,7 +352,13 @@ def is_present(self):
         """
         Helper method to see if this is a delete vs. update
         """
-        if self.identifier and (self.attachment_src == self.attachment_from is None):
-            return False
-        else:
-            return True
+        # FIXME in its previous form this always returned True
```
[the remark this answers is not in the document]
Comment (millerdev) [reply corehq/form_processor/abstract_models.py]: Restored in d004a43f72c62587c82909604fdf1f5d40a0bb99

File: corehq/form_processor/abstract_models.py [hunk 8]
```diff
@@ -354,7 +352,13 @@ def is_present(self):
         """
         Helper method to see if this is a delete vs. update
```
[the remark this answers is not in the document]
Comment (millerdev) [reply corehq/form_processor/abstract_models.py]: d004a43f72c62587c82909604fdf1f5d40a0bb99

File: corehq/ex-submodules/casexml/apps/case/templates/case/partials/case_attachments.html (hunk 2 above)
Comment (millerdev) [reply corehq/ex-submodules/casexml/apps/case/templates/case/partials/case_attachments.html]: 5745015c2a29e64860e48f90e46903a9b957746b

**Comment by millerdev** [comment]
I'm not sure how to fix [this test failure](https://travis-ci.org/dimagi/commcare-hq/jobs/399773279), @emord or @snopoke can you shed any light on that?
```
FAIL: testapps.test_elasticsearch.tests.test_prod_es_indices:ProdIndexManagementTest.test_prod_config
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/vendor/lib/python2.7/site-packages/django/test/utils.py", line 384, in inner
    return func(*args, **kwargs)
  File "/mnt/commcare-hq-ro/testapps/test_elasticsearch/tests/test_prod_es_indices.py", line 31, in test_prod_config
    self.assertEqual(EXPECTED_PROD_INDICES, found_prod_indices)
AssertionError: Lists differ: [{u'alias': u'case_search', u'... != [{'alias': u'case_search', u'm...

First differing element 1:
{u'alias': u'hqapps', u'meta': {u'settings': {u'number_of_replicas': 0, u'analysis': {u'analyzer': {u'default': {u'filter': [u'lowercase'], u'type': u'custom', u'tokenizer': u'whitespace'}}}}}, u'type': u'app', u'index': u'test_hqapps_2017-05-22_1426'}
{'alias': u'report_cases', u'meta': {u'settings': {u'number_of_replicas': 0, u'analysis': {u'analyzer': {u'default': {u'filter': [u'lowercase'], u'type': u'custom', u'tokenizer': u'whitespace'}, u'sortable_exact': {u'filter': [u'lowercase'], u'type': u'custom', u'tokenizer': u'keyword'}}}}}, 'type': u'report_case', 'index': u'test_dm/case-attachment-refactor-8f7cba681695730b5f639f430441f148984615b8'}
Diff is 10669 characters long. Set self.maxDiff to None to see it.
```

**Comment by millerdev** [comment]
> I'm not sure how to fix this test failure...

resolved

File: corehq/ex-submodules/casexml/apps/case/sharedmodels.py [hunk 9]
```diff
@@ -79,8 +79,26 @@ class CommCareCaseAttachment(LooselyEqualDocumentSchema, CaseAttachmentMixin, Un
...
+    @property
+    def is_present(self):
+        """
+        Helper method to see if this is a delete vs. update
+
+        NOTE this is related to but reversed logic from
+        `casexml.apps.case.xml.parser.CaseAttachment.is_delete`.
+        """
+        # Should this only use `attachment_src` since that is the
+        # only field checked by `from_case_index_update`?
```
Comment (snopoke) [inline corehq/ex-submodules/casexml/apps/case/sharedmodels.py]: According to the [casexml spec](https://github.com/dimagi/commcare-core/wiki/casexml20) all of `src`, `from` and `name` should be empty for a deletion. I would say that as long as `src` and `from` are empty we should consider it a deletion.

Note that there is a case in the spec (not implemented by us) where `src` could be empty: if `from = 'inline'`

File: corehq/ex-submodules/casexml/apps/case/sharedmodels.py [hunk 10]
```diff
@@ -79,8 +79,26 @@ class CommCareCaseAttachment(LooselyEqualDocumentSchema, CaseAttachmentMixin, Un
...
+            self.identifier and
+            self.attachment_src is self.attachment_from is None
+        )
+
     @classmethod
     def from_case_index_update(cls, attachment):
+        # yet another is_deleted check? but different from the others;
+        # this one only looks at attachment_src instead of some combination
+        # of attachment_src, attachment_from, and attachment_name
         if attachment.attachment_src:
```
Comment (snopoke) [inline corehq/ex-submodules/casexml/apps/case/sharedmodels.py]: Guessing this is OK in our implementation since we don't support `inline` attachments. Would be better to standardize though.

File: corehq/ex-submodules/casexml/apps/case/xml/parser.py [hunk 11]
```diff
@@ -171,6 +171,23 @@ def __init__(self, identifier, attachment_src, attachment_from, attachment_name)
...
+        """
+        Helper method to see if this is a delete vs. update
+
+        This property is named differently from
+        `casexml.apps.case.sharedmodels.CommCareCaseAttachment.is_present`
+        to disambiguate the interface of that class from this one since
+        they are not the same but are used in very similar contexts.
+        """
+        # FIXME slight disparity between this and CaseAttachmentAction.from_v2
+        # comment: "this is a deletion if all three ... are null"
```
Comment (snopoke) [inline corehq/ex-submodules/casexml/apps/case/xml/parser.py]: The comment you're referring to doesn't include `identifier`. If there is no `identifier` then we wouldn't know which attachment to delete.

File: corehq/form_processor/abstract_models.py [hunk 12]
```diff
@@ -288,12 +288,13 @@ def get_properties_in_api_format(self):
...
+        return {
+            name: {
                 'url': self.get_attachment_server_url(att.identifier),
-                'mime': att.attachment_from
-            }) for name, att in self.case_attachments.items()
-        ])
+                'content_type': att.content_type,
+            }
+            for name, att in self.case_attachments.items()
+        }
```
Comment (snopoke) [inline corehq/form_processor/abstract_models.py]: :+1: thanks

File: corehq/pillows/mappings/reportcase_mapping.py [hunk 13]
```diff
@@ -4,9 +4,9 @@
 from pillowtop.es_utils import ElasticsearchIndexInfo
 
 
-REPORT_CASE_INDEX = es_index("report_cases_czei39du507m9mmpqk3y01x72a3ux4p0")
+REPORT_CASE_INDEX = es_index("dm-case-attachment-refactor-8f7cba6")
```
Comment (snopoke) [inline corehq/pillows/mappings/reportcase_mapping.py]: This index name should still be called `report_cases_<something>` so that it's clear which index it is. The convention with the other indexes is to append the date. You can always update the `comment` field if you want to add more details.

**Review by snopoke — COMMENTED** [review]
A few small comments but otherwise looks good.

File: corehq/ex-submodules/casexml/apps/case/xml/parser.py (hunk 11 above)
Comment (millerdev) [reply corehq/ex-submodules/casexml/apps/case/xml/parser.py]: Yes, `identifier` makes sense, it's the fact that this only checks for null `attachment_src` and `attachment_from` while the other comment references `attachment_name` as well. Other places doing deleted checks are also inconsistent with both this code and that comment. It's hard to follow because there is not one consistent way to check for deleted status. It's not even clear if that's what's being checked in some places, one has to guess. However, this is old code and I'm hesitant to change it now because that could have unintended consequences—who knows if there is code elsewhere that has come to depend on these inconsistencies? Hopefully not.

Comment (snopoke) [reply corehq/ex-submodules/casexml/apps/case/xml/parser.py]: I'd be very pro standardizing this to `delete = identifier and not (attachment_src or attachment_from)`

**Comment by millerdev** [comment]
✅

File: corehq/ex-submodules/casexml/apps/case/xml/parser.py (hunk 11 above)
Comment (millerdev) [reply corehq/ex-submodules/casexml/apps/case/xml/parser.py]: I'm curious why `identifier` has to be considered for this check? I understand why we need it to actually *do* the deletion. If `src` and `from` are blank then it's a deletion; not having an identifier means we don't know what to delete. That would be a data issue, not an indication of whether it was a deletion or not.

Comment (snopoke) [reply corehq/ex-submodules/casexml/apps/case/xml/parser.py]: True, we could validate that all attachments have an identifier elsewhere though since the identifier comes from the XML element name I don't think it's possible for it to be blank at this point.

Comment (millerdev) [reply corehq/ex-submodules/casexml/apps/case/xml/parser.py]: https://github.com/dimagi/commcare-hq/pull/21106/commits/8f61e0a46f14a664a4cfb9381784e87e690ed0ec

**Comment by millerdev** [comment]
@snopoke any reason you're holding off on approving this?
